File: Java/TransportServer/src/main/java/transport/server/TransportServicesImpl.java
package transport.server;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.mindrot.jbcrypt.BCrypt;
import transport.model.Flight;
import transport.model.Ticket;
import transport.model.User;
import transport.persistence.repositoryBD.FlightRepoBD;
import transport.persistence.repositoryBD.TicketRepoBD;
import transport.persistence.repositoryBD.UserRepoBD;
import transport.services.ITransportObserver;
import transport.services.ITransportServices;
import transport.services.TransportException;


import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.stream.Collectors;

public class TransportServicesImpl implements ITransportServices {
    private static final Logger logger = LogManager.getLogger();
    private FlightRepoBD flightRepository;
    private TicketRepoBD ticketRepository;
    private UserRepoBD userRepository;

    private Map<String, ITransportObserver> loggedClient;
    private final Map<ITransportObserver, Boolean> observers = new ConcurrentHashMap<>();

    static private TransportServicesImpl instance;

    public static TransportServicesImpl getInstance(FlightRepoBD flightRepository, TicketRepoBD ticketRepository, UserRepoBD userRepository) {
        if (instance == null) {
            instance = new TransportServicesImpl(flightRepository, ticketRepository, userRepository);
        }
        return instance;
    }

    private TransportServicesImpl(FlightRepoBD flightRepository, TicketRepoBD ticketRepository, UserRepoBD userRepository) {
        this.flightRepository = flightRepository;
        this.ticketRepository = ticketRepository;
        this.userRepository = userRepository;
        loggedClient = new HashMap<>();
    }

    public List<String> searchFlights(String destination, LocalDate departureDate) throws TransportException {
        try{
            return flightRepository.searchByDestinationAndDate(destination, departureDate).stream()
                    .map(flight -> flight.getDestination() + " "+ flight.getDepartureTime().toLocalDate() + " " + flight.getDepartureTime().toLocalTime() +
                            " Seats: " + flight.getAvailableSeats())
                    .collect(Collectors.toList());
        }
        catch (Exception e){
            throw new TransportException(e.getMessage());
        }

    }

    @Override
    public List<Flight> getAllFlights() throws TransportException {
        try {
            System.out.println("Getting all flights... ");
            Iterable<Flight> iterableFlights = flightRepository.findAll();
            List<Flight> flightList = new ArrayList<>();
            iterableFlights.forEach(flightList::add);
            return flightList;
        }catch (Exception e){
            throw new TransportException("Ceva in repo nu a mers bine/n/n/n/n!!!!!!!!!!!!!!!!!!!!!!!!!!!");
        }
    }

    public synchronized boolean buyTicket(String selectedFlight, int seats, List<String> passengers) {

        System.out.println("BuyTicket Service!!!");
        String destination = selectedFlight.split(" ")[0];
        String departure = selectedFlight.split(" ")[1] + " " + selectedFlight.split(" ")[2];;

        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
        LocalDateTime departureDateTime = LocalDateTime.parse(departure, formatter);

        System.out.println(destination);
        System.out.println(departureDateTime);

        Flight flight = flightRepository.searchByDestinationAndDateTime(destination, departureDateTime).stream().findFirst().orElse(null);


        if (flight == null || flight.getAvailableSeats() < seats) {
            return false;
        }

        Ticket ticket = new Ticket(passengers, flight.getId(), seats);
        ticketRepository.add(ticket);
        updateFlightSeats(flight.getId(), -seats);


        ExecutorService executor = Executors.newSingleThreadExecutor();
        executor.submit(() -> {
            for (ITransportObserver observer : observers.keySet()) {
                try {
                    observer.reservationAdded("NOTIFICARE");
                    System.out.println("Notificare trimisă către observer: " + observer);
                } catch (TransportException e) {
                    System.err.println("Eroare la notificare: " + e.getMessage());
                }
            }
        });
        executor.shutdown();



        return true;
    }

    private synchronized void updateFlightSeats(Integer flightId, int seatChange) {
        Flight flight = flightRepository.find(flightId);
        if (flight != null) {
            int newSeats = flight.getAvailableSeats() + seatChange;
            if (newSeats < 0) {
                logger.warn("Attempted to reduce seats below zero for flight {}", flightId);
                return;
            }
            flight.setAvailableSeats(newSeats);
            flightRepository.update(flightId, flight);
        }
    }

    public synchronized void addUser(User user) {
        logger.info("Adding user: {}", user);

        String hashedPassword = hashPassword(user.getPassword());
        user.setPassword(hashedPassword);

        userRepository.add(user);
    }

    public User findUser(String username) {
        return userRepository.find(username);
    }


    public synchronized void login(String username, String password, ITransportObserver client) throws TransportException {
        User user = userRepository.find(username);
        if (user != null && checkPassword(password, user.getPassword())) {
            loggedClient.put(username, client);
            addObserver(client);
        } else {
            throw new TransportException("Invalid username or password");
        }
    }

    @Override
    public void logout(String username, ITransportObserver client) throws TransportException {
        ITransportObserver localClient = loggedClient.remove(username);
        removeObserver(localClient);
    }

    public synchronized String hashPassword(String password) {
        return BCrypt.hashpw(password, BCrypt.gensalt());
    }

    private synchronized boolean checkPassword(String password, String hashedPassword) {
        return BCrypt.checkpw(password, hashedPassword);
    }


    public void addObserver(ITransportObserver observer) {
        observers.put(observer, true);
    }

    public void removeObserver(ITransportObserver observer) {
        observers.remove(observer);
    }


}

File: C#/TransportServer/transport/server/TransportServicesImpl.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using log4net;

using proiect.repository.repositoryDB;
using Transport.Services;
using TransportModel.transport.model;
using TransportServices.transport.services;

public class TransportServicesImpl : ITransportServices
    {
        private static readonly ILog logger = LogManager.GetLogger(typeof(TransportServicesImpl));
        private readonly FlightRepoDB flightRepository;
        private readonly TicketRepoDB ticketRepository;
        private readonly UserRepoDB userRepository;
        private User loginUser;
        private readonly IDictionary<string, ITransportObserver> loggedClients;

        public TransportServicesImpl(FlightRepoDB flightRepository, TicketRepoDB ticketRepository, UserRepoDB userRepository)
        {
            this.flightRepository = flightRepository;
            this.ticketRepository = ticketRepository;
            this.userRepository = userRepository;
            this.loggedClients = new Dictionary<string, ITransportObserver>();
        }


        public void AddFlight(Flight flight)
        {
            logger.Info($"Adding flight: {flight}");
            flightRepository.Add(flight);
        }


        public Flight FindFlight(int id)
        {
            lock (this)
            {
                return flightRepository.Find(id);
            }
        }


        public List<Flight> GetAllFlights()
        {
            lock (this)
            {
                return flightRepository.FindAll().ToList();
            }
        }


        public void UpdateFlight(int id, Flight flight)
        {
            lock (this)
            {
                flightRepository.Update(id, flight);
            }
        }


        public void DeleteFlight(int id)
        {
            lock (this)
            {
                flightRepository.Delete(id);
            }
        }


        public List<Flight> SearchFlights(string destination, DateTime departureDate)
        {
            lock (this)
            {
                return flightRepository.SearchByDestinationAndDate(destination, departureDate);
            }
        }


        public bool BuyTicket(string destination, int requiredseats, int seats, List<string> passengers)
        {
            lock (this)
            {
                var flight = flightRepository.SearchBySeatsAndDestination(destination, requiredseats).FirstOrDefault();

                Console.WriteLine("flight: " + flight);

                if (flight == null || flight.AvailableSeats < seats)
                {
                    Console.WriteLine("nu o gasit repo-ul zborul");
                    return false;
                }

                var ticket = new Ticket(seats, passengers, flight.Id);
                ticketRepository.Add(ticket);
                UpdateFlightSeats(flight.Id, -seats);
                
                
                foreach (ITransportObserver client in loggedClients.Values)
                {
                    try
                    {
                        Task.Run(() => client.ReservationAdded("da"));
                    }
                    catch (TransportException e)
                    {
                        Console.Error.WriteLine("Error notifying client in TransportServicesImpl: " + e);
                    }
                }
                
                
                return true;
            }
        }


        public Ticket FindTicket(int id)
        {
            lock (this)
            {
                return ticketRepository.Find(id);
            }
        }


        public IEnumerable<Ticket> GetAllTickets()
        {
            lock (this)
            {
                return ticketRepository.FindAll();
            }
        }
        

 
        private void UpdateFlightSeats(int flightId, int seatChange)
        {
            lock (this)
            {
                var flight = flightRepository.Find(flightId);
                if (flight != null)
                {
                    int newSeats = flight.AvailableSeats + seatChange;
                    if (newSeats < 0)
                    {
                        logger.Warn($"Attempted to reduce seats below zero for flight {flightId}");
                        return;
                    }

                    flight.AvailableSeats = newSeats;
                    flightRepository.Update(flightId, flight);
                }
            }
        }
        
        public void AddUser(User user)
        {
            lock (this)
            {
                logger.Info($"Adding user: {user}");
                string hashedPassword = HashPassword(user.Password);
                user.Password = hashedPassword;
                userRepository.Add(user);
            }
        }


        public User FindUser(string username)
        {
            lock (this)
            {
                return userRepository.Find(username);
            }
        }


        public IEnumerable<User> GetAllUsers()
        {
            lock (this)
            {
                return userRepository.FindAll();
            }
        }


        public bool Login(string username, string password, ITransportObserver observer)
        {
            lock (this)
            {
                Console.WriteLine("In serverImp: try login");
                var user = userRepository.Find(username);
                Console.WriteLine(username);
                if (user != null && CheckPassword(password, user.Password))
                {
                    Console.WriteLine("In serverImp: Login successful");
                    loggedClients[user.Username] = observer;
                    return true;
                }
                else
                {
                    return false;
                }
            }
        }
        

        public string HashPassword(string password)
        {
            lock (this)
            {
                return BCrypt.Net.BCrypt.HashPassword(password);
            }
        }


        private bool CheckPassword(string password, string hashedPassword)
        {
            lock (this)
            {
                return BCrypt.Net.BCrypt.Verify(password, hashedPassword);
            }
        }
    }
